# 异步解决方案及详细介绍
## 异步的概念
打个比方：现在执行一个事情，但这个事情不是现在立刻就执行，而是过一段时间再执行。过一段时间之后才能得知结果
## 异步解决方案
* 回调函数实现（例如 setTimeout通过回调函数实现）
* ES6 中使用promise，Generator 到 async / await 
### 一、回调函数
* 优点： 解决了同步的问题
* 缺点： 如何一个事情需要依赖另一个事情的执行结果，则需要嵌套执行，层层嵌套是相当繁琐的。导致“回调地狱”，不好调试，语义不明确
```javascript
ajax('XXX1', () => {
    // callback 函数体
    ajax('XXX2', () => {
        // callback 函数体
        ajax('XXX3', () => {
            // callback 函数体
        })
    })
})
```

### 二、 Promise
* 优点：promise 可以通过.then()来执行下一个操作，语义化，便于调试，解决了回调地狱的问题
* 缺点：
  * promise 同样也是链式调用，调用多个也将会比较长
  * 无法取消Promise，一旦新建它就会立即执行，无法中途取消。
  * 如果不设置回调函数，Promise内部抛出的错误，不会反应到外部。
  * 当处于pending状态时，无法得知目前进展到哪一个阶段（刚刚开始还是即将完成）。
  
#### Promise基本语法
Promise对象代表一个异步操作，有三种状态：pending（进行中）、fulfilled（已成功）和rejected（已失败）

``` javascript
const promise = new Promise(function(resolve, reject) {
  // ... some code

  if (/* 异步操作成功 */){
    resolve(value);
  } else {
    reject(error);
  }
});

promise.then(function(value) {
  // success
}, function(error) {
  // failure
});
```
Promise 新建后就会立即执行。Promise本身是同步的立即执行函数， 当在 executor 中执行 resolve 或者 reject 的时候, 此时是异步操作， 会先执行 then/catch 等，当主栈完成后，才会去调用 resolve/reject 中存放的方法执行。

``` javascript
let promise = new Promise(function(resolve, reject) {
  console.log('Promise');
  resolve();
});

promise.then(function() {
  console.log('resolved.');
});

console.log('Hi!');

// Promise
// Hi!
// resolved
```
#### 方法1、Promise.all() 

```javascript
const p = Promise.all([p1, p2, p3]);
```
Promise.all()方法接受一个数组作为参数，p1、p2、p3都是 Promise 实例，如果不是，就会先调用下面讲到的Promise.resolve方法，将参数转为 Promise 实例，再进一步处理。另外，Promise.all()方法的参数可以不是数组，但**必须具有 Iterator 接口，且返回的每个成员都是 Promise 实例**。
* **只有p1、p2、p3的状态都变成fulfilled，p的状态才会变成fulfilled**，此时**p1、p2、p3的返回值组成一个数组，传递给p的回调函数**。
* **只要p1、p2、p3之中有一个被rejected，p的状态就变成rejected**，此时**第一个被reject的实例的返回值，会传递给p的回调函数**。 --- **不能保证p1,p2,p3都有最后结果状态，无法确定所有请求都结束。**
* 如果作为参数的 Promise 实例，自己定义了catch方法，那么它一旦被rejected，并不会触发Promise.all()的catch方法

```javascript
const p1 = new Promise((resolve, reject) => {
  resolve('hello');
})
.then(result => result)
.catch(e => e);

const p2 = new Promise((resolve, reject) => {
  throw new Error('报错了');
})
.then(result => result)
.catch(e => e);

Promise.all([p1, p2])
.then(result => console.log(result))
.catch(e => console.log(e));
// ["hello", Error: 报错了]
```
上面代码中，p1会resolved，p2首先会rejected，但是p2有自己的catch方法，该方法返回的是一个新的 Promise 实例，p2指向的实际上是这个实例。该实例执行完catch方法后，也会变成resolved，导致Promise.all()方法参数里面的两个实例都会resolved，因此会调用then方法指定的回调函数，而不会调用catch方法指定的回调函数。

如果p2没有自己的catch方法，就会调用Promise.all()的catch方法。
```javascript
const p1 = new Promise((resolve, reject) => {
  resolve('hello');
})
.then(result => result);

const p2 = new Promise((resolve, reject) => {
  throw new Error('报错了');
})
.then(result => result);

Promise.all([p1, p2])
.then(result => console.log(result))
.catch(e => console.log(e));
// Error: 报错了
```
#### 方法2、Promise.allSettled()
Promise.allSettled()方法接受一组 Promise 实例作为参数，
**只有等到所有这些参数实例都返回结果，不管是fulfilled还是rejected，包装实例才会结束。**
该方法返回的新的 Promise 实例，一旦结束，**状态总是fulfilled，不会变成rejected** 。状态变成fulfilled后，Promise 的监听函数接收到的**参数是一个数组**，每个成员对应一个传入Promise.allSettled()的 Promise 实例。
```javascript
const resolved = Promise.resolve(42);
const rejected = Promise.reject(-1);

const allSettledPromise = Promise.allSettled([resolved, rejected]);

allSettledPromise.then(function (results) {
  console.log(results);
});
// [
//    { status: 'fulfilled', value: 42 },
//    { status: 'rejected', reason: -1 }
// ]
```
Promise.allSettled()的返回值allSettledPromise，状态只可能变成fulfilled。它的监听函数接收到的参数是数组results。该数组的每个成员都是一个对象，对应传入Promise.allSettled()的两个 Promise 实例。每个对象都有status属性，该属性的值只可能是字符串fulfilled或字符串rejected。fulfilled时，对象有value属性，rejected时有reason属性，对应两种状态的返回值。

**我们不关心异步操作的结果，只关心这些操作有没有结束。这时，Promise.allSettled()方法就很有用，Promise.all()无法确定所有请求都结束，想要达到这个目的，写起来很麻烦**

#### 方法3、Promise.race()

```javascript
const p = Promise.race([p1, p2, p3]);
```
Promise.race()方法的参数与Promise.all()方法一样
**只要p1、p2、p3之中有一个实例率先改变状态，p的状态就跟着改变。那个率先改变的 Promise 实例的返回值，就传递给p的回调函数。**

#### Promise.any() 
Promise.any()方法接受一组 Promise 实例作为参数，包装成一个新的 Promise 实例。
* **只要参数实例有一个变成fulfilled状态，包装实例就会变成fulfilled状态；**
* **如果所有参数实例都变成rejected状态，包装实例就会变成rejected状态。**
* Promise.any()跟Promise.race()方法很像，只有一点不同，就是不会因为某个 Promise 变成rejected状态而结束。

#### 方法4、Promise.prototype.catch() 
Promise.prototype.catch() 方法是.then(null, rejection)或.then(undefined, rejection)的别名，用于指定发生错误时的回调函数。catch()方法**返回的还是一个 Promise 对象**，因此后面还可以接着调用then()方法。
**注意点**
* Promise 对象的错误具有“冒泡”性质，会一直向后传递，直到被捕获为止。也就是说，错误总是会被下一个catch语句捕获
* 一般来说，不要在then()方法里面定义 Reject 状态的回调函数（即then的第二个参数），总是使用catch方法。
```javascript
// bad
promise
  .then(function(data) {
    // success
  }, function(err) {
    // error
  });

// good
promise
  .then(function(data) { //cb
    // success
  })
  .catch(function(err) {
    // error
  });
```
#### 方法5、Promise.prototype.finally()

Promise.prototype.finally()用于指定不管 Promise 对象最后状态如何，都会执行的操作

#### 方法6、Promise.resolve()

有时需要将现有对象转为 Promise 对象，Promise.resolve()方法就起到这个作用。
```javascript
Promise.resolve('foo')
// 等价于
new Promise(resolve => resolve('foo'))

```
#### 方法7、Promise.reject()

Promise.reject(reason)方法也会返回一个新的 Promise 实例，该实例的状态为rejected。
Promise.reject()方法的参数，会原封不动地作为reject的理由，变成后续方法的参数。这一点与Promise.resolve方法不一致。
```javascript
const p = Promise.reject('出错了');
// 等同于
const p = new Promise((resolve, reject) => reject('出错了'))

p.then(null, function (s) {
  console.log(s)
});
// 出错了
```

### 三、 async 
async函数的返回值是 Promise 对象，可以使用then方法添加回调函数。当函数执行的时候，一旦遇到await就会先返回，等到异步操作完成，再接着执行函数体内后面的语句。await命令后面是一个 Promise 对象，返回该对象的结果，如果不是 Promise 对象，会调用 Promise.resolve(值) 转化为Promise 对象。

```javascript
// 今日头条面试题
async function async1() {
    console.log('async1 start')
    await async2()
    console.log('async1 end')
}
async function async2() {
    console.log('async2')
}
console.log('script start')
setTimeout(function () {
    console.log('settimeout')
})
async1()
new Promise(function (resolve) {
    console.log('promise1')
    resolve()
}).then(function () {
    console.log('promise2')
})
console.log('script end')

 // script start
 // async1 start
 // async2
 // promise1
 // script end
 // async1 end
 // promise2
 // settimeout
```
（1）async1，async2都是函数定义，先同步任务，打印script start

（2）setTimeout 先挂起。

（3）执行async1()，打印 async1 start，调用async2()，打印async2，先返回，将它后面的代码 `console.log('async1 end')` push 进微任务队列中。

（4）因为 Promise 会立即执行，所以打印出 promise1，将.then 里面的代码 ` console.log('promise2') ` push进微任务队列中

（5）打印 script end 。同步任务执行完，接下来从微任务队列中取。

（6）打印 async1 end

（7）打印 promise2。此时微任务队列为空，则去宏任务队列中取。

（8）到时间(浏览器最小时间4ms)，将setTimeout中回调函数push 进宏任务队列中

（9）执行回调函数 打印 settimeout

#### Promise 与 async 区别
* promise执行异步的方式是 : .then() .then 链式调用
* async 是 await funcA;  await funcB;  将异步变成同步的写法,本质还是异步的。
